### PR TITLE
[Feature][Connector V2] expose configurable options in IoTDB

### PR DESCRIPTION
**Pull request by john8628**
## Purpose of this pull request
https://github.com/apache/incubator-seatunnel/issues/3386

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.apache.seatunnel.connectors.seatunnel.iotdb.sink;
+
+import static org.apache.seatunnel.connectors.seatunnel.iotdb.config.CommonConfig.NODE_URLS;
```
Comment (CalvinKirs) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java:20]: You need add the License header

Comment (john8628) [reply seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java:20]: ok,thanks

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 2]
```diff
@@ -35,28 +38,28 @@
...
-    public static final String MAX_THRIFT_FRAME_SIZE = "max_thrift_frame_size";
-    public static final String ZONE_ID = "zone_id";
-    public static final String ENABLE_RPC_COMPRESSION = "enable_rpc_compression";
-    public static final String CONNECTION_TIMEOUT_IN_MS = "connection_timeout_in_ms";
-
     private static final int DEFAULT_BATCH_SIZE = 1024;
 
+    public static final Option<String> KEY_TIMESTAMP = Options.key("key_timestamp").stringType().noDefaultValue().withDescription("key timestamp");
+    public static final Option<String> KEY_DEVICE = Options.key("key_device").stringType().noDefaultValue().withDescription("key device");
+    public static final Option<String> KEY_MEASUREMENT_FIELDS = Options.key("key_measurement_fields").stringType().noDefaultValue().withDescription("key measurement fields");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]:
```suggestion
    public static final Option<String> KEY_MEASUREMENT_FIELDS = Options.key("key_measurement_fields").listType().noDefaultValue().withDescription("key measurement fields");
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 3]
```diff
@@ -35,28 +38,28 @@
...
+    public static final Option<String> STORAGE_GROUP = Options.key("storage_group").stringType().noDefaultValue().withDescription("store group");
+    public static final Option<Integer> BATCH_SIZE =  Options.key("batch_size").intType().defaultValue(DEFAULT_BATCH_SIZE).withDescription("batch size");
+    public static final Option<String> BATCH_INTERVAL_MS = Options.key("batch_interval_ms").stringType().noDefaultValue().withDescription("batch interval ms");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java:48]:
```suggestion
    public static final Option<String> BATCH_INTERVAL_MS = Options.key("batch_interval_ms").intType().noDefaultValue().withDescription("batch interval ms");
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 4]
```diff
@@ -35,28 +38,28 @@
...
+    public static final Option<Integer> MAX_RETRIES = Options.key("max_retries").intType().noDefaultValue();
+    public static final Option<String> RETRY_BACKOFF_MULTIPLIER_MS = Options.key("retry_backoff_multiplier_ms").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]:
```suggestion
    public static final Option<String> RETRY_BACKOFF_MULTIPLIER_MS = Options.key("retry_backoff_multiplier_ms").intType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 5]
```diff
@@ -35,28 +38,28 @@
...
+    public static final Option<String> MAX_RETRY_BACKOFF_MS = Options.key("max_retry_backoff_ms").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]:
```suggestion
    public static final Option<String> MAX_RETRY_BACKOFF_MS = Options.key("max_retry_backoff_ms"). intType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 6]
```diff
@@ -35,28 +38,28 @@
...
+    public static final Option<String> DEFAULT_THRIFT_BUFFER_SIZE = Options.key("default_thrift_buffer_size").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]:
```suggestion
    public static final Option<String> DEFAULT_THRIFT_BUFFER_SIZE = Options.key("default_thrift_buffer_size"). intType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 7]
```diff
@@ -35,28 +38,28 @@
...
+    public static final Option<Integer> MAX_THRIFT_FRAME_SIZE = Options.key("max_thrift_frame_size").intType().noDefaultValue();
+    public static final Option<String> ZONE_ID = Options.key("zone_id").stringType().noDefaultValue();
+    public static final Option<String> ENABLE_RPC_COMPRESSION = Options.key("enable_rpc_compression").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]:
```suggestion
    public static final Option<String> ENABLE_RPC_COMPRESSION = Options.key("enable_rpc_compression").booleanType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java [hunk 8]
```diff
@@ -35,28 +38,28 @@
...
+    public static final Option<String> CONNECTION_TIMEOUT_IN_MS = Options.key("connection_timeout_in_ms").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]:
```suggestion
    public static final Option<String> CONNECTION_TIMEOUT_IN_MS = Options.key("connection_timeout_in_ms").intType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java [hunk 9]
```diff
@@ -25,78 +28,78 @@
...
      * The host of the IotDB server.
      */
-    public static final String HOST = "host";
+    public static final Option<String> HOST = Options.key("host").stringType().noDefaultValue().withDescription("host");
 
     /*
      * The port of the IotDB server.
      */
-    public static final String PORT = "port";
+    public static final Option<String> PORT = Options.key("port").stringType().noDefaultValue().withDescription("port");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java]:
```suggestion
    public static final Option<String> PORT = Options.key("port").intType().noDefaultValue().withDescription("port");
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java [hunk 10]
```diff
@@ -25,78 +28,78 @@
...
-    public static final String NODE_URLS = "node_urls";
+    public static final Option<String> NODE_URLS = Options.key("node_urls").stringType().noDefaultValue().withDescription("node urls");
 
     /*---------------------- other configurations -------------------------*/
 
     /**
      * Fetches the next batch of data from the source.
      */
-    public static final String FETCH_SIZE = "fetch_size";
+    public static final Option<String> FETCH_SIZE = Options.key("fetch_size").stringType().noDefaultValue().withDescription("fetch size");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java]:
```suggestion
    public static final Option<String> FETCH_SIZE = Options.key("fetch_size").intType().noDefaultValue().withDescription("fetch size");
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java [hunk 11]
```diff
@@ -25,78 +28,78 @@
...
      * thrift max frame size
      */
-    public static final String THRIFT_MAX_FRAME_SIZE = "thrift_max_frame_size";
+    public static final Option<Integer> THRIFT_MAX_FRAME_SIZE = Options.key("thrift_max_frame_size").intType().noDefaultValue().withDescription("thrift max frame size of iot db");
 
     /**
      * cassandra default buffer size
      */
-    public static final String ENABLE_CACHE_LEADER = "enable_cache_leader";
+    public static final Option<String> ENABLE_CACHE_LEADER = Options.key("enable_cache_leader").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java]:
```suggestion
    public static final Option<String> ENABLE_CACHE_LEADER = Options.key("enable_cache_leader").booleanType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java [hunk 12]
```diff
@@ -25,78 +28,78 @@
...
      * Version represents the SQL semantic version used by the client, which is used to be compatible with the SQL semantics of 0.12 when upgrading 0.13. The possible values are: V_0_12, V_0_13.
      */
-    public static final String VERSION = "version";
+    public static final Option<String> VERSION = Options.key("version").stringType().noDefaultValue();
 
     /**
      * Query lower bound of the time range to be read.
      */
-    public static final String LOWER_BOUND = "lower_bound";
+    public static final Option<String> LOWER_BOUND = Options.key("lower_bound").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java]:
```suggestion
    public static final Option<String> LOWER_BOUND = Options.key("lower_bound").longType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java [hunk 13]
```diff
@@ -25,78 +28,78 @@
...
 
     /**
      * Query upper bound of the time range to be read.
      */
-    public static final String UPPER_BOUND = "upper_bound";
+    public static final Option<String> UPPER_BOUND = Options.key("upper_bound").stringType().noDefaultValue();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SourceConfig.java]:
```suggestion
    public static final Option<String> UPPER_BOUND = Options.key("upper_bound").longType().noDefaultValue();
```

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+    @Override
+    public String factoryIdentifier() {
+        return "IoTDB";
+    }
+
+    @Override
+    public OptionRule optionRule() {
+        return OptionRule.builder()
+                .required(NODE_URLS, USERNAME, PASSWORD, SinkConfig.KEY_DEVICE)
+                .optional().build();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java]:
```suggestion
                .optional(KEY_TIMESTAMP, xxx, .....).build();
```

add all optional keys

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceFactory.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+public class IoTDBSourceFactory implements TableSourceFactory{
+    @Override
+    public String factoryIdentifier() {
+        return "IoTDB";
+    }
+
+    @Override
+    public OptionRule optionRule() {
+        return OptionRule.builder()
+                .required(NODE_URLS, USERNAME, PASSWORD, SourceConfig.SQL).build();
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceFactory.java]: add all optional keys

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java (hunk 2 above)
Comment (john8628) [reply seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/config/SinkConfig.java]: OK ,got it . thanks

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceFactory.java (hunk 15 above)
Comment (john8628) [reply seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceFactory.java]: OK ,got it . thanks

File: seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java (hunk 14 above)
Comment (john8628) [reply seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/sink/IoTDBSinkFactory.java]: OK ,got it . thanks

**Comment by hailin0** [comment]
@john8628 
Add if null check to IoTDBSourceReader#close

image

**Comment by john8628** [comment]
> @john8628 Add if null check to IoTDBSourceReader#close
> 
> image

OK ;

**Comment by hailin0** [comment]
using `conf.get(xxx.getKey())` of all options


https://github.com/apache/incubator-seatunnel/actions/runs/3480966388/jobs/5821511988
image

**Review by hailin0 — APPROVED** [review]

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**